sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -221,44 +255,112 @@ public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string stream
+                        {
+                            Task runningTask = runningTasks[i];
+                            if (!runningTask.IsCompleted)
+                            {
+                                continue;
+                            }
+
+                            await runningTask.ConfigureAwait(false);
+                            runningTasks.RemoveAt(i);
+                            i--;
+                        }
                     }
                 }
+
+                // Wait for all the remaining blocks to finish uploading
+                await Task.WhenAll(runningTasks).ConfigureAwait(false);
+
+                // Process all errors after tasks are done to determine status
+                // Will run on a single thread
+                foreach (Task<RunningTask<T>> task in runningTasks)
+                {
+                    // Go through errors from each task and if we have an error, add error from response into errors list which represents the errors from all the batches
+                    if (task.Result.Error != null)
+                        errors.Add(task.Result.Error);
+                }
             }
-            catch (Exception e)
+            catch (Exception ex)
             {
-                scope.Failed(e);
+                scope.Failed(ex);
                 throw;
             }
 
-            UploadLogsResult finalResult = new UploadLogsResult(errors, Status(logEntries, errors));
+            // Calculate the status using the helper method Status
+            UploadLogsResult finalResult = new UploadLogsResult(errors, Status(logs, errors));
             return Response.FromValue(finalResult, response);
         }
 
+        private async Task<RunningTask<T>> CommitBatchListSyncOrAsync<T>(BatchedLogs<T> batch, string ruleId, string streamName, bool async, CancellationToken cancellationToken)
+        {
+            UploadLogsError error;
+            RequestContext requestContext = GenerateRequestContext(cancellationToken);
+            Response response = null;
+
+            using HttpMessage message = CreateUploadRequest(ruleId, streamName, batch.LogsData, "gzip", requestContext);
+
+            if (async)
+            {
+                response = await _pipeline.ProcessMessageAsync(message, requestContext, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                response = _pipeline.ProcessMessage(message, requestContext, cancellationToken);
+            }
+
+            if (response.Status != 204) // if any error is thrown log it
+            {
+                RequestFailedException requestFailedException = new RequestFailedException(response);
+                ResponseError responseError = new ResponseError(requestFailedException.ErrorCode, requestFailedException.Message);
+                List<Object> objectLogs = new List<Object>((IEnumerable<object>)batch.LogsList);
```

Comment:
Why is this a list of `object`? Shouldn't it be BinaryData?

Reply:
Related to this - should the [FailedLogs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.Ingestion/src/Models/UploadLogsError.cs#L25) property be an `IReadOnlyList<BinaryData>`?

Reply:
yes - I think I had it as more generic before but I will update

Reply:
specifying also reduces the number of casts